slogr/handler.go

```diff
@@ -0,0 +1,89 @@
+//go:build go1.21
+// +build go1.21
+
+/*
+Copyright 2023 The logr Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package slogr
+
+import (
+	"context"
+	"log/slog"
+
+	"github.com/go-logr/logr"
+)
+
+// NewSlogHandler returns a slog.Handler which logs through an arbitrary
+// logr.Logger.
+func NewSlogHandler(logger logr.Logger) slog.Handler {
+	return &logrHandler{
+		// Account for extra wrapper functions.
+		logger: logger.WithCallDepth(3),
+	}
+}
+
+// logrHandler implements slog.Handler in terms of a logr.Logger.
+type logrHandler struct {
+	logger logr.Logger
+}
+
+func (h logrHandler) Enabled(_ context.Context, level slog.Level) bool {
+	return h.logger.V(int(-level)).Enabled()
+}
+
+func (h logrHandler) Handle(_ context.Context, record slog.Record) error {
+	//FIXME: I don't know what to do with record.Time or record.PC.  Neither of
+	// them map to logr.
+	args := recordToKV(record)
+	if record.Level >= slog.LevelError {
+		h.logger.Error(nil, record.Message, args...)
+	} else {
+		h.logger.V(int(-record.Level)).Info(record.Message, args...)
```

Comment:
Consider dividing by 4, so:
```
Info -> V(0)
Debug -> V(1)
Warn -> V(-1)
```
I don't know if that's better.

Reply:
A direct mapping is better because Kubernetes uses V(4) for debug messages, which corresponds nicely with slog.LevelDebug = -4.

Reply:
Agree - I don't see why we need to modify it - am I missing some trap?